File: LeetCode75_Artur/DP - 1D/1137_N-th_Tribonacci_Number.cpp
/*
 https://leetcode.com/problems/n-th-tribonacci-number/?envType=study-plan-v2&envId=leetcode-75
*/

#include <iostream>
#include <vector>

using namespace std;

class Solution
{
public:
    int tribonacci(int n)
    {
        if( n == 0 )
            return 0;
        else if( n <= 2 )
            return 1;

        auto dp = vector<int>(n + 1, 0);
        dp[0] = 0;
        dp[1] = 1;
        dp[2] = 1;

        for(int i = 3; i <= n; ++i)
            dp[i] = dp[i - 1] + dp[i - 2] + dp[i - 3];

        return dp[n];
    }
};


int main()
{
    Solution solution;

    cout << solution.tribonacci(4) << endl;
    cout << solution.tribonacci(7) << endl;
    cout << solution.tribonacci(10) << endl;
    cout << solution.tribonacci(15) << endl;
    cout << solution.tribonacci(20) << endl;
    cout << solution.tribonacci(25) << endl;
}

/*
 Main idea:
 Динаміка, ідентична із задачею числами Фібоначчі,
 тільки цього разу додаються 3 попередні числа, а не
 2.

 Складність алгоритму: O(n)
*/

File: LeetCode75_Artur/DP - 1D/1137_N-th_Tribonacci_Number.py
"""
https://leetcode.com/problems/n-th-tribonacci-number/?envType=study-plan-v2&envId=leetcode-75
"""

from typing import Optional, List

class Solution:
    def tribonacci(self, n: int) -> int:
        if n == 0:
            return 0
        
        n_0 = 0
        n_1 = 1
        n_2 = 1
        
        for _ in range(3, n + 1):
            n_curr = n_0 + n_1 + n_2
            n_0 = n_1
            n_1 = n_2
            n_2 = n_curr
            
        return n_2
    

if (__name__ == '__main__'):
    solution = Solution()
    
    print(solution.tribonacci(2))
    print(solution.tribonacci(4))
    print(solution.tribonacci(5))
    print(solution.tribonacci(10))
    print(solution.tribonacci(15))
    print(solution.tribonacci(20))
    
    
    